File: Group-Video/OpenVideoCall-iOS-Objective-C/OpenVideoCall/MessageViewController.m
//
//  MessageViewController.m
//  OpenVideoCall
//
//  Created by CavanSu on 2019/6/6.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "MessageViewController.h"
#import "MessageCell.h"

@interface MessageViewController () <UITableViewDelegate, UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITableView *messageTableView;
@property (nonatomic, strong) NSMutableArray *list;
@end

@implementation MessageViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.messageTableView.rowHeight = UITableViewAutomaticDimension;
    self.messageTableView.estimatedRowHeight = 24;
}

- (void)appendInfo:(NSString *)text {
    Message *message = [Message messageWithText:text type:MessageTypeInfo];
    [self appendMessage:message];
}

- (void)appendError:(NSString *)text {
    Message *message = [Message messageWithText:text type:MessageTypeError];
    [self appendMessage:message];
}

#pragma mark - Private
- (void)appendMessage:(Message *)message {
    [self.list insertObject:message atIndex:0];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:0];
    [self.messageTableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationLeft];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.list.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MessageCell *cell =  [tableView dequeueReusableCellWithIdentifier:@"MessageCell"];
    Message *message = self.list[indexPath.row];
    [cell setMessage:message];
    return cell;
}

- (NSMutableArray *)list {
    if (!_list) {
        _list = [NSMutableArray array];
    }
    return _list;
}
@end

File: Group-Video/OpenVideoCall-iOS/OpenVideoCall/MessageViewController.swift
//
//  MessageViewController.swift
//  OpenVideoCall
//
//  Created by GongYuhua on 16/8/15.
//  Copyright © 2016年 Agora. All rights reserved.
//

import UIKit

class MessageViewController: UIViewController {
    
    @IBOutlet weak var messageTableView: UITableView!
    
    private var messageList = [Message]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        messageTableView.rowHeight = UITableView.automaticDimension
        messageTableView.estimatedRowHeight = 24
    }
    
    func append(info text: String) {
        let message = Message(text: text, type: .info)
        append(message: message)
    }
    
    func append(alert text: String) {
        let message = Message(text: text, type: .alert)
        append(message: message)
    }
}

private extension MessageViewController {
    func append(message: Message) {
        messageList.append(message)
        
        var deleted: Message?
        if messageList.count > 20 {
            deleted = messageList.removeFirst()
        }
        
        updateMessageTable(with: deleted)
    }
    
    func updateMessageTable(with deleted: Message?) {
        guard let tableView = messageTableView else {
            return
        }
        
        let insertIndexPath = IndexPath(row: messageList.count - 1, section: 0)
        
        tableView.beginUpdates()
        if deleted != nil {
            tableView.deleteRows(at: [IndexPath(row: 0, section: 0)], with: .none)
        }
        tableView.insertRows(at: [insertIndexPath], with: .none)
        tableView.endUpdates()
        
        tableView.scrollToRow(at: insertIndexPath, at: .bottom, animated: false)
    }
}

extension MessageViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return messageList.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "messageCell", for: indexPath) as! MessageCell
        let message = messageList[indexPath.row]
        cell.set(with: message)
        return cell
    }
}
